### ISSUE TITLE
Add a column for Parameters to BlitzWho

### ISSUE THREAD

**Issue by gdoddsy** [BrentOzarULTD/SQL-Server-First-Responder-Kit#2591]
**Is your feature request related to a problem? Please describe.**
I want to see the parameters that were compiled into a plan in BlitzWho results. Ideally I can also see the current running parameter.

**Describe the solution you'd like**
A new column (or 2) in BlitzWho that shows the compiled and runtime parameters for a query. Something like:
**i {int} : Compiled: (0) Running:(1)**

**Describe alternatives you've considered**
The data is in the query plan, so you can always just open it and read it

**Are you ready to build the code for the feature?**
I've already started

**Comment by gdoddsy** [comment]
@BrentOzar - a couple of questions/problems:

- Do you have a better suggestion for the formatting of the column?
- When adding a column to the output table, do I just drop it into the table create statement in the order that the output would be? And then do I do a check like the JoinKey does (line 260ish in the current dev branch) to add it to existing tables?
- When shredding the XML I'm getting access violation errors on the live query plan but not the query plan from the cache. I can get around it by dropping the output into a temp table and then working on that, but I don't think that's an ideal solution. I'll keep trying to work out an option, but can you think of anything obvious about why the XML for the live plan can't be shredded, but the cached plan can?

**Comment by gdoddsy** [comment]
I think I've found the problem with shredding the XML and can avoid SQL crashing my query. I can get either the parameters from the cached query plan or from the live query plan, but not both.

Am I right in saying that the Actual plan should be the same as the cached plan, just with the actuals up to the point of capturing it? I.e. for the sake of this issue, the compiled parameters on the cached and live plan would be the same?

I'm thinking for the output I'll just have a "Cached" or "Live" label on the row so you know where to look:
Cached: @i {int} : (0)
Live: @i {int} : (0) (Actual: (0))

Sound ok? Still not clear on adding columns to the table, I'll give it a crack and if you don't like it you can fix that bit up. I'll hopefully find some time later this week to fix it all up and send in a pull request.

**Comment by BrentOzar** [comment]
Yeah, the actual & cached plan should be the same here. Good job! I love it. Thanks sir!

**Comment by gdoddsy** [comment]
@BrentOzar - I discovered that there are differences between the cached and actual plans - the actual only shows the currently running statement of the query, the cached plan shows all statements compiled together. I've added two columns, one for the cached so you can see all params cached, and one for the actual so you can see what is running live right now. Hope that makes sense, let me know any problems.

I've only tested this on 2019 on the StackOverflow database. I've put in code to protect the live query plans in older versions, but I don't have a server to test it out on. It looks the same as the other protections in place.

**Comment by BrentOzar** [comment]
Looks good! Merging into the dev branch, will be in the November release with credit to you in the release notes. Good work on adding it to the output table too!

**Comment by BrentOzar** [comment]
OOO, we have a problem. I should have tested this on your branch before I took the pull request, hahaha.

How to repro - in one window, run this:

```
CREATE OR ALTER PROC ##usp_Counter @Test1 INT, @Test2 INT AS
BEGIN
	SELECT COUNT(*) FROM StackOverflow.dbo.Posts WITH (INDEX = 1) WHERE @Test1 = @Test2;
	SELECT COUNT(*) FROM StackOverflow.dbo.Posts WITH (INDEX = 1) WHERE @Test2 = @Test1;
END
GO

EXEC ##usp_Counter 1, 1;
```

Then in another window:

`EXEC ##usp_Counter 2, 2;`

Then finally in another window, while those two are executing, run sp_BlitzWho, and on SQL 2019 CU8 I get a stack dump:

```
Location:	 tmpilb.cpp:1063
Expression:	 0 == m_cLength || m_cLength == cLength
SPID:		 62
Process ID:	 3592
Msg 3624, Level 20, State 1, Line 24
A system assertion check has failed. Check the SQL Server error log for details. Typically, an assertion failure is caused by a software bug or data corruption. To check for database corruption, consider running DBCC CHECKDB. If you agreed to send dumps to Microsoft during setup, a mini dump will be sent to Microsoft. An update might be available from Microsoft in the latest Service Pack or in a Hotfix from Technical Support.
Msg 596, Level 21, State 1, Line 0
Cannot continue the execution because the session is in the kill state.
Msg 0, Level 20, State 0, Line 0
A severe error occurred on the current command.  The results, if any, should be discarded.
```

Ouch! I've reverted the pull request for now. I would love to help dig into this, but I'm slammed right now - if you want to go for it, let me know when you think it's in a position to test again, and I'll give it another shot.

**Comment by gdoddsy** [comment]
AHHH! That was the problem I had which I thought I'd fixed. I didn't test it with more than 1 query - oops.

Keep it open, I'll give it another go, but it might be a couple of weeks till I get back to this one.

**Comment by gdoddsy** [comment]
So I've finally found a fix for this, using your test case above, I now see this output:

![image](https://user-images.githubusercontent.com/25063374/106375468-6435be80-63e0-11eb-8f55-51ca454fd6fd.png)


To do that, I'll need to access `sys.dm_exec_query_statistics_xml(s.session_id)` a bit differently. It appears that the stack dump happens when you have 2 columns trying to read query_plan, in our case it's to display the query plan and then to strip the parameters. I still can't find anyone else hitting this error and I don't have a budget to raise a ticket with MS, but I can prove the behavior.

I can fix it in one of two ways:

1. Add a CTE at the top of the select (after inserting the data into `@blocked`) that will look something like:
```
WITH LiveQueryPlans AS(
	SELECT s.session_id, dm_exec_query_statistics_xml.query_plan
	FROM		sys.dm_exec_sessions				  AS s
	CROSS APPLY sys.dm_exec_query_statistics_xml(s.session_id)
)
```

and then left join to it; OR
2. Do the same but into a table variable like @Blocked does.

I could also hide all of this behind a parameter, i.e. if you don't want to see the actual parameters then it will keep the current functionality (you can still get compiled params just fine), if you do want to see them then it'll add the CTE/Table Variable.

I'm a little concerned by the CTE option, as I think I have had the same stack dump as above once. I'm not 100% sure, as I may have had an old version of the proc running, and I haven't been able to recreate it. I'm pretty confident that dumping it all in the table variable will be fine (MS Docs has that exact query). But then I have a temp table, and I know how you feel about those.

I should be able to sort it out this week, so appreciate any thoughts you might have. If I don't hear from you because Iceland is too beautiful or you've been eaten by a [...well it appears there are no animals there that kill you, you should come down here to Australia...], then I'll assume the table variable is a good way to go.

**Comment by BrentOzar** [comment]
I like your line of thinking on both:
* I'd add a parameter like ShowActualParameters, default to 0. We can always change the default to 1 later if we use the bejeezus out of it in the wild and we don't see issues.
* I'd also do the table variable. I think it's the safest.

Good work, sir!

**Comment by BrentOzar** [comment]
Nice job! Merged into the dev branch, will be in the April release with credit to you in the release notes.

**Comment by BrentOzar** [comment]
Also, in case anyone's looking for an example of how the output looks:
<img width="680" alt="Screen Shot 2021-03-24 at 5 37 38 AM" src="https://user-images.githubusercontent.com/245462/112261960-b5d22b00-8c64-11eb-9cbe-906f42ff44dd.png">

**Comment by gdoddsy** [comment]
Found a problem with it already, but not a show stopper. On one server I'm getting the Cached Parameter Info showing data twice, I think it's to do with LAST_QUERY_PLAN_STATS = ON. The XML in the cached query plan has 2 sets of parameters can I can't tell which one to pick, so it's just picking both.
No idea if I'm going to be able to fix that (or even if I care enough to fix it) as I can't see anything obvious in the XML to identify which is which.

**Comment by gdoddsy** [comment]
I'm getting Access Violation errors again in SQL 2019 due to this bit of work.

> Msg 596, Level 21, State 1, Line 0
> Cannot continue the execution because the session is in the kill state.
> Msg 0, Level 20, State 0, Line 0
> A severe error occurred on the current command.  The results, if any, should be discarded.

It's not happening on all servers, but it is happening for one of them, I'm not sure what makes that server special. I'm hopeful CU12 will fix the problem with this bug fix: https://support.microsoft.com/en-gb/topic/kb5004524-cumulative-update-12-for-sql-server-2019-45b2d82a-c7d0-4eb8-aa17-d4bad4059987#bkmk_14089504.

Not sure if BlitzWho needs a flag to turn off the query plan for these cases. Thoughts?

**Comment by BrentOzar** [comment]
Ouch! Yeah, then let's add a new parameter for it, and default it to off. When there's a possibility of an error to be generated when we're just gathering nice-to-have stuff, I'd rather default it to off, and have people opt into it.

**Comment by gdoddsy** [comment]
Added a pull request
